package spaceopera;

// all message types
enum message_types {
	
	// auth step
	
	mt_hello = 0;
	mt_hello_reply = 1;

	// universe management step
	
	mt_get_universes = 10;
	mt_get_universes_reply = 11;
	mt_connect_to_universe = 12;
	mt_connect_to_universe_reply = 13;
	
	// universe
	
	mt_universe_state_feed = 20;
	mt_universe_control = 21;
	mt_universe_control_reply = 22;
}

// header - fixed size (5 bytes for each , sent before each message
message header {
	required sfixed32 message_len = 1;
	required sfixed32 message_type = 2;
	// this is breaking layer, I'm not proud. Always 0 for aync notification sent from server to client
	required sfixed32 request_id = 3;
	
}

//////////////////////////////
// authentication step

// C-> S, clinet identification
message hello {
	required string user_name = 1;
	optional string password = 2;
}

// S->, identification ok or error
message hello_reply {
	required bool ok = 1;
	optional string message = 2;
}

///////////////////////////////
// universe management step

// C -> s, request list of all universes
message get_universes {
}

message get_universes_reply {
	repeated string universe_name = 1;
}

message connect_to_universe {
	required string universe_name = 1;
}

message connect_to_universe_reply {
	required bool ok = 1;
	optional string message = 2;
}

////////////////////////////////
// universe session

message universe_state_feed {
	required double current_time = 1;
	optional double time_scale = 2;
	repeated string removed_objects = 3;
	repeated object_description added_objects = 4;
}

	message object_description {
		required string name = 1;
	}

message universe_control {
	optional double new_time_scale = 1;
}

message universe_control_reply {
	required bool ok = 1;
	optional string message = 2;
}







